backend/src/main/java/atlantbh/restaurants/services/LocationService.java

```diff
@@ -0,0 +1,45 @@
+package atlantbh.restaurants.services;
+
+import atlantbh.restaurants.models.Location;
+import atlantbh.restaurants.models.filters.LocationFilterBuilder;
+import atlantbh.restaurants.models.sortkeys.LocationSortKeys;
+import atlantbh.restaurants.repositories.LocationRepository;
+import org.hibernate.Criteria;
+import org.hibernate.criterion.Restrictions;
+import org.hibernate.query.Query;
+import org.springframework.stereotype.Service;
+
+import java.util.*;
+
+@Service
+public class LocationService extends BaseService<Location, LocationSortKeys, LocationFilterBuilder, LocationRepository> {
+
+    // TODO Change method to use Criteria instead of native query
+    public List<Location> findAllCitiesForCountry(String country) {
+        Query query = repository.getSession().createNativeQuery("SELECT *  FROM Location l WHERE l.id IN " +
+                                                                    "(SELECT MIN(loc.id) FROM Location loc " +
+                                                                    "GROUP BY loc.city) and l.country = :country", Location.class);
+        query.setParameter("country", country);
+        return query.getResultList();
+
+    }
+    // TODO Change method to use Criteria instead of native query
+    public List<Location> findAllCountires() {
+       Query query = repository.getSession().createNativeQuery("SELECT *  FROM Location l WHERE l.id IN (SELECT MIN(loc.id) FROM Location loc GROUP BY loc.country)", Location.class);
+        return query.getResultList();
+    }
+
+    public Collection<Location> getTopLocations(Integer numOfLocations) throws Exception {
+
+        // outdated method
+        // TODO retrieving top locations with count criteria
+        ArrayList<Location> locations = new ArrayList<>(repository.findAll());
```

Comment:
no, we will not fetch all and do some sorting in memory, it should be on db level..

Reply:
Fixed 8c841a20e489a31c66d177c8682339f9d4f19de0